rfcs/SubschemaComposability.md

```diff
@@ -0,0 +1,239 @@
+   (gateways, gateway schemas, etc.), the execution of a request against a
+   composite schema will yield request(s) against spec-compliant (or even
+   non-spec compliant) GraphQL subservices that are then merged into a single
+   request.
+2. `Single-service composite schema execution`: Often, especially when utilizing
+   "build-time" composition, a single service executes the composite schema
+   without delegation to remote services tied to each subschema. Subschemas
+   exist for the purpose of code organization, but only the composite schema
+   exists at the time of execution.
+
+The below RFC is meant to be useful across all varieties of schema composition.
+Because all forms of schema composition require a composite schema, we can speak
+of "composability" without referring to the timing of composition or the type of
+execution. However, practically speaking, the method of composite schema
+execution matters significantly (much more than the timing of composition). For
+example, if a request against the composite schema must ultimately be translated
+to request(s) against the still extant source subschemas, then the composition
+must be -- in some sense -- reversed during execution, which is not always
+possible, and introduces significant complexity.
+
+The below RFC examines composability separately for the above two forms of
+composite schema execution.
+
+# Techniques for Maintaining Composability
+
+## `Namespacing`
+
+A set of potentially "overlapping" types in subschemas may be considered to have
+been "namespaced" if they have been adjusted in some way so as to longer be
+overlapping. For example, identically named types in different subschemas may be
+namespaced by utilizing some form of metadata to map the identically named types
+to different types in the composite schema. Or, when metadata denotes types as
+overlapping in the first instance, namespacing may be accomplished simply by
+adjusting the existing metadata or via additional metadata. Finally, a new
+namespace construct may be introduced into the GraphQL specification itself such
+that, for example, multiple, different types with the same name could be
+organized into separate namespaces without conflict. In any of these situations,
+the "namespaced" GraphQL elements no longer overlap and no longer _require_
+composition; there is therefore no impediment to composability.
+
+## `Transforms`
+
+Namespacing is one way in which subschemas may be adjusted prior to composition.
+More broadly, schemas may be transformed in many ways, many of which may impact
+composition. Types, fields, enum values, etc., for example, can be renamed,
+removed, or added, sometimes allowing a set of previously non composable
+subschemas to become composable. When execution involves requests against the
+subschemas as remote GraphQL services, the subschema transformation may also
+require runtime transformation of the requests to the subservices and/or the
+subservice results.
+
```

Comment:
One composition technique I found useful for my team is employed by [Nautilus](https://github.com/nautilus/gateway). It composes subschema object types if and only if they opt-in. They opt-in by implementing a shared interface with only an `id: ID!` field. It appears to be unique in that it isn't specifying directives or special syntax.

The spec for "global object identification" lends itself well to schema merging for my team's needs. This solved many of our schema composition problems in a very simple way, i.e. subschemas did not need to go out of their way to opt-in to composition for objects.

This may even be in a separate category from Namespaces or Transforms. Curious what you think there.

Reply:
Personally, I think it's a great technique when it's available, but it's not always an option.

For example, in my team's usecase [at Kensho](https://blog.kensho.com/database-agnostic-querying-is-unavoidable-at-scale-18895f6df2f0), we could not have used this approach because we didn't "own" the underlying data sources, and many of them didn't even have a single (or any) primary key that we could reference. There was data with composite keys, and data that was generated on demand and had no meaningful natural key of any kind whatsoever. Additionally, for the most part the underlying services weren't aware they were being queried through GraphQL and a metadata-driven federated execution system, and could not meaningfully have opted into federation.

I completely agree that opt-in is great and should be used whenever possible. However, mandating opt-in behavior in the standard would disallow use cases like the one I described and linked to, which would be a shame and would likely lead to more non-standard extensions being developed and used.

Reply:
@obi1kenobi You make some good points about the tradeoffs of using `Node` as an opt-in. This was mainly a suggestion for an additional "in the wild" strategy for merging, below namespacing and transforms.